darshan-runtime/lib/darshan-ldms.c

```diff
@@ -0,0 +1,249 @@
+#include <netdb.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+/* Check for LDMS libraries if Darshan is built --with-ldms */
+#ifdef HAVE_LDMS
+#include <ldms/ldms.h>
+#include <ldms/ldmsd_stream.h>
+#include <ovis_util/util.h>
+#include "ovis_json/ovis_json.h"
+#include "darshan-ldms.h"
+
+struct darshanConnector dC = {
+     .ldms_darsh = NULL,
+     .ldms_lib = 0,
+     };
+
+ldms_t ldms_g;
+static void event_cb(ldms_t x, ldms_xprt_event_t e, void *cb_arg)
+{
+        switch (e->type) {
+        case LDMS_XPRT_EVENT_CONNECTED:
+                sem_post(&dC.conn_sem);
+                dC.conn_status = 0;
+                break;
+        case LDMS_XPRT_EVENT_REJECTED:
+                ldms_xprt_put(x);
+                dC.conn_status = ECONNREFUSED;
+                break;
+        case LDMS_XPRT_EVENT_DISCONNECTED:
+                ldms_xprt_put(x);
+                dC.conn_status = ENOTCONN;
+                break;
+        case LDMS_XPRT_EVENT_ERROR:
+                dC.conn_status = ECONNREFUSED;
+                break;
+        case LDMS_XPRT_EVENT_RECV:
+                sem_post(&dC.recv_sem);
+                break;
+        case LDMS_XPRT_EVENT_SEND_COMPLETE:
+                break;
+        default:
+                printf("Received invalid event type %d\n", e->type);
+        }
+}
+
+#define SLURM_NOTIFY_TIMEOUT 5
+ldms_t setup_connection(const char *xprt, const char *host,
+                        const char *port, const char *auth)
+{
+        char hostname[PATH_MAX];
+        const char *timeout = "5";
+        int rc;
+        struct timespec ts;
+
+        if (!host) {
+                if (0 == gethostname(hostname, sizeof(hostname)))
+                        host = hostname;
+        }
+        if (!timeout) {
+                ts.tv_sec = time(NULL) + 5;
+                ts.tv_nsec = 0;
+        } else {
+                int to = atoi(timeout);
+                if (to <= 0)
+                        to = 5;
+                ts.tv_sec = time(NULL) + to;
+                ts.tv_nsec = 0;
+        }
+
+        ldms_g = ldms_xprt_new_with_auth(xprt, NULL, auth, NULL);
+        if (!ldms_g) {
+                printf("Error %d creating the '%s' transport\n",
+                       errno, xprt);
+                return NULL;
+        }
+
+        sem_init(&dC.recv_sem, 1, 0);
+        sem_init(&dC.conn_sem, 1, 0);
+
+        rc = ldms_xprt_connect_by_name(ldms_g, host, port, event_cb, NULL);
+        if (rc) {
+                printf("Error %d connecting to %s:%s\n",
+                       rc, host, port);
+                return NULL;
+        }
+        sem_timedwait(&dC.conn_sem, &ts);
+        if (dC.conn_status)
+                return NULL;
+        return ldms_g;
+}
+
+void darshan_ldms_connector_initialize()
+{
+    if (!getenv("DARSHAN_LDMS_STREAM"))
+        dC.env_ldms_stream = "darshanConnector";
+
+    /* Set flags for various LDMS environment variables */
+    if (getenv("DXT_ENABLE_LDMS"))
```

Comment:
I think we should get rid of any leftover references to DXT as far as LDMS integrations go. DXT is just a module that captures finer-grained tracing details for POSIX/MPI-IO modules, but I don't think there's any need for LDMS to be aware of it. LDMS should be able to get all the information it needs using the integrations into MPI-IO and POSIX modules directly.